drivers/flash/flash_esp32.c

```diff
@@ -69,21 +70,49 @@ static inline void flash_esp32_sem_give(const struct device *dev)
 
 #endif /* CONFIG_MULTITHREADING */
 
+#include <zephyr/kernel.h>
+#include <zephyr/logging/log.h>
+#include <zephyr/sys/util.h>
+#include <stdint.h>
+#include <string.h>
+
 static int flash_esp32_read(const struct device *dev, off_t address, void *buffer, size_t length)
 {
 	int ret = 0;
 
-	flash_esp32_sem_take(dev);
+#ifdef CONFIG_MCUBOOT
+	/* ensure everything is 4-byte aligned */
+	size_t aligned_length = ROUND_UP(length, 4);
+	off_t aligned_address = ROUND_DOWN(address, 4);
+	size_t address_offset = address - aligned_address;
+	uint32_t temp_buf[aligned_length / 4];
+
 	if (!esp_flash_encryption_enabled()) {
-		ret = esp_flash_read(NULL, buffer, address, length);
+		ret = esp_rom_flash_read(aligned_address, temp_buf, aligned_length, false);
 	} else {
+		ret = esp_rom_flash_read(aligned_address, temp_buf, aligned_length, true);
+	}
+
+	memcpy((void *)buffer, ((uint8_t *)temp_buf) + address_offset, length);
```

Comment:
@sylvioalves This doesn't look correct.

Let's consider an example with `address = 3` and `length = 7`.

It would give the following values:

```
aligned_length = 8;
aligned_address = 0;
address_offset = 3;
temp_buf[2];
```

The bytes returned would not contain all desired bytes and we would even read beyond the size of `temp_buf` (marked with `?` below):

```
flash content           | 0 1 2 3 | 4 5 6 7 | 8 9 A B |
bytes wanted            | - - - 3 | 4 5 6 7 | 8 9 - - |
bytes read to temp_buf  | 0 1 2 3 | 4 5 6 7 |   
bytes returned          | 3 4 5 6 | 7 ? ? - |   
```

In addition to that, the entire `temp_buf` is allocated on the stack, which is not ideal IMO. This could easily lead to a stack overflow if large amounts of data are read/written.

Reply:
Sure, let me re-check and work this out.